### ISSUE TITLE
Streaming - Video Lagging / Stutter

### ISSUE THREAD

**Issue by arrmo** [fHDHR/fHDHR#256]
OK, was thinking this was just me - then I heard "feedback" from my daughter ... 🤣. She's seeing the same issue. Let me try to explain.

I was trying some streaming, to VLC (playlist / m3u). Seemed to be great, but after a period of time I started to gradually see more and more packet errors => blocky video, lagging, stutter ... not sure 100% what to call it. I put it aside, figured I'd get back to it, to try to figure out what was causing it.

Last night, my daughter reported the same issue, using Plex. So I restarted Plex ... no change. Then I tried restarting fHDHD ... fixed it immediately. But I just heard from her, it lasted about an hour, then started to happen again - the same as I saw, it gradually starts to degrade ... then the fix is to restart fHDHR. Some sort of memory leak perhaps? Or caching related somehow? I'm open to thoughts!

Have you seen similar things? I'm not seeing much in the logs, that help to pinpoint what is going on - or not any recurring errors / issues at least.

I'm wondering about trying buffering off? Suggestions? I did try changing the read size, make it a multiple of 3984, but no delta. And using ffmpeg or direct, both seem to be the same as well.

Thanks!

**Comment by deathbybandaid** [comment]
Not experiencing that at all with my streams

**Comment by arrmo** [comment]
Figured it may be specific to this case 😆. So here is something a bit odd - not sure if this helps or not, but things I'm trying to get to the bottom of it. Ran 2 cases (simultaneously! On 2 tuners), ~ 200k frames run,
1. Manually tuned to a channel, then had VLC directly open the device (/dev/ctn91xx). 5 dropped frames.
2. VLC in a second window, but this time through fHDHR (ceton = direct). 70 dropped frames.

Not earth shattering, and not what I was seeing before. So then, in parallel I fired up Plex, through fHDHR (another tuner) ... and all heck broke loose! Within the next ~ 10k frames, the second case above => 290 dropped frames, but none on the first case ... so not the tuner itself, nor the Linux driver - agreed? And in Plex ... yep, burping and farting 🤣. I stopped Plex, and ... dropped frames slowed way back down again (case 2 above). So seems to be tied somehow to streaming to Plex? BTW, in VLC, I also saw a large number of Dropped (Input/Read), seems to be when Plex streaming is in place (I think). Not sure if that helps or not, but it may.

Things to try? Perhaps disabling all buffering? Just thinking out loud.

Thanks!

**Comment by arrmo** [comment]
Actually, another thought on this ... is the "outgoing" streaming different - to Plex vs. m3u? I could try the same approach in VLC (RTP?) ... if that's possible? Just trying to help narrow down where it's at.

Thanks!

**Comment by deathbybandaid** [comment]
All the streaming from fHDHR exits from the api/tuners endpoint

**Comment by arrmo** [comment]
Let me do some digging - you lost me there ... LMAO! Just trying to see if I can mimic the Plex connection, but direct to VLC.

Thanks!

**Comment by deathbybandaid** [comment]
the HDHR plugin, the m3u, everything just directs the client (plex, vlc) to `api/tuners?method=stream&channel=XXX`

**Comment by arrmo** [comment]
Got it, thanks! Seems the channel is (sometimes?) a UUID? Or at least the m3u looks like that 😉

**Comment by arrmo** [comment]
OK, it seems that Plex is not really part of the equation. I set up 3 VLC instances,
1. Direct from the device - this one stays clean, really no issues.
2. m3u, same channel as 1.
3. m3u, another channel

2 stays clean, until 3 starts ... then lots of drops, on 2 and 3 (only), and these errors (ran VLC from the CLI, to watch for issues),
```
[00007fb8e0002170] ts demux error: libdvbpsi error (PSI decoder): TS discontinuity (received 9, expected 8) for PID 0
[00007fb8e0002170] ts demux error: libdvbpsi error (PSI decoder): TS discontinuity (received 0, expected 15) for PID 1070
[00007fb8e0002170] ts demux error: libdvbpsi error (PSI decoder): TS discontinuity (received 0, expected 15) for PID 4770
[00007fb8e0002170] ts demux error: libdvbpsi error (PSI decoder): TS discontinuity (received 0, expected 15) for PID 4870
[00007fb8e0002170] ts demux error: libdvbpsi error (PSI decoder): TS discontinuity (received 11, expected 10) for PID 0
[00007fb8e0002170] ts demux error: libdvbpsi error (PSI decoder): TS discontinuity (received 8, expected 7) for PID 4770
[00007fb8e0002170] ts demux error: libdvbpsi error (PSI decoder): TS discontinuity (received 8, expected 7) for PID 4870
[00007fb8e0002170] ts demux error: libdvbpsi error (PSI decoder): TS discontinuity (received 6, expected 5) for PID 0
[00007fb8e0002170] ts demux error: libdvbpsi error (PSI decoder): TS discontinuity (received 10, expected 8) for PID 0
[00007fb8e0002170] ts demux error: libdvbpsi error (PSI decoder): TS discontinuity (received 12, expected 11) for PID 0
[00007fb8e0002170] ts demux error: libdvbpsi error (PSI decoder): TS discontinuity (received 12, expected 10) for PID 4610
[00007fb8e0002170] ts demux error: libdvbpsi error (PSI decoder): TS discontinuity (received 12, expected 11) for PID 4770
[00007fb8e0002170] ts demux error: libdvbpsi error (PSI decoder): TS discontinuity (received 12, expected 11) for PID 4870
[00007fb8e0002170] ts demux error: libdvbpsi error (PSI decoder): TS discontinuity (received 12, expected 10) for PID 0
[00007fb8e0002170] ts demux error: libdvbpsi error (PSI decoder): TS discontinuity (received 1, expected 14) for PID 0
[00007fb8e0002170] ts demux error: libdvbpsi error (PSI decoder): TS discontinuity (received 10, expected 9) for PID 1050
[00007fb8e0002170] ts demux error: libdvbpsi error (PSI decoder): TS discontinuity (received 11, expected 10) for PID 4600
[00007fb8e0002170] ts demux error: libdvbpsi error (PSI decoder): TS discontinuity (received 5, expected 3) for PID 4610
[00007fb8e0002170] ts demux error: libdvbpsi error (PSI decoder): TS discontinuity (received 5, expected 3) for PID 1070
[00007fb8e0002170] ts demux error: libdvbpsi error (PSI decoder): TS discontinuity (received 5, expected 3) for PID 4770
```

Then I stopped 3 -> 2 clean again. So it seems some sort of interaction within fHDHR? Thoughts?

Thanks!!

**Comment by deathbybandaid** [comment]
Channels are most easily identified by uuid, but there is logic for handling numbers

**Comment by arrmo** [comment]
Thanks! OK, trying another tact ... thinking that buffering may be the issue, and RTP may help / fix it? RTP wasn't an option before, but with the firmware updates => I can try that now. So I modified the code (locally), to try RTP, but I get this error (saw that one before ... have you seen it?).
```
[2022-02-19 16:54:57,714] p2990286 {/mnt/ProgSSD/fHDHR/fHDHR/streammanager/stream_obj.py:118} - DEBUG - Attempting to gather stream information for ceton channel 593 Golf HD Golf HD.
[2022-02-19 16:54:57,798] p2990286 {/mnt/ProgSSD/fHDHR/plugins/fHDHR_plugin_origin_ceton/origin/__init__.py:73} - INFO - Selected Ceton tuner#: 1
[2022-02-19 16:54:57,798] p2990286 {/mnt/ProgSSD/fHDHR/plugins/fHDHR_plugin_origin_ceton/origin/__init__.py:85} - INFO - Ceton tuner 1 to be started
[2022-02-19 16:54:58,130] p2990286 {/mnt/ProgSSD/fHDHR/plugins/fHDHR_plugin_origin_ceton/origin/__init__.py:193} - INFO - Preparing Ceton tuner 1 on port: 41708
[2022-02-19 16:54:58,409] p2990286 {/mnt/ProgSSD/fHDHR/plugins/fHDHR_plugin_origin_ceton/origin/__init__.py:206} - INFO - Initiate PCIe direct streaming, channel 593 from Ceton tuner#: 1 
[2022-02-19 16:54:58,410] p2990286 {/mnt/ProgSSD/fHDHR/plugins/fHDHR_plugin_origin_ceton/origin/__init__.py:209} - INFO - streamurl: rtp://192.168.200.2:8000
[2022-02-19 16:54:58,411] p2990286 {/mnt/ProgSSD/fHDHR/fHDHR/streammanager/stream_obj.py:94} - INFO - Attempting to Select an available tuner for this stream.
[2022-02-19 16:54:58,412] p2990286 {/mnt/ProgSSD/fHDHR/fHDHR/device/tuners/tuner.py:76} - INFO - Tuner #1 Acquired.
[2022-02-19 16:54:58,413] p2990286 {/mnt/ProgSSD/fHDHR/fHDHR/streammanager/stream_obj.py:102} - INFO - ceton Tuner #1 to be used for stream.
[2022-02-19 16:54:58,414] p2990286 {/mnt/ProgSSD/fHDHR/fHDHR/streammanager/stream_obj.py:104} - INFO - Preparing Stream...
[2022-02-19 16:54:58,414] p2990286 {/mnt/ProgSSD/fHDHR/fHDHR/streammanager/stream/__init__.py:51} - INFO - Stream Method Detected as RTP/s.
[2022-02-19 16:54:58,415] p2990286 {/mnt/ProgSSD/fHDHR/fHDHR/streammanager/stream/direct_rtp_stream.py:21} - INFO - Attempting to create socket to listen on.
[2022-02-19 16:54:58,416] p2990286 {/mnt/ProgSSD/fHDHR/fHDHR/streammanager/stream/direct_rtp_stream.py:33} - INFO - Created TCP socket at 192.168.2.64:44843.
[2022-02-19 16:54:58,417] p2990286 {/mnt/ProgSSD/fHDHR/fHDHR/streammanager/stream/direct_rtp_stream.py:40} - INFO - Created UDP socket at 192.168.2.64:56624.
[2022-02-19 16:54:58,418] p2990286 {/mnt/ProgSSD/fHDHR/fHDHR/streammanager/stream/direct_rtp_stream.py:68} - INFO - Closing UDP socket at 192.168.2.64:56624.
[2022-02-19 16:54:58,418] p2990286 {/mnt/ProgSSD/fHDHR/fHDHR/streammanager/stream/direct_rtp_stream.py:70} - INFO - Closing TCP socket at 192.168.2.64:44843.
[2022-02-19 16:54:58,419] p2990286 {/mnt/ProgSSD/fHDHR/fHDHR_web/api/tuners.py:38} - ERROR - Unable to Stream: TunerError: 806 - Tune Failed: Could Not Create Socket: 'username'
[2022-02-19 16:54:58,420] p2990286 {/mnt/ProgSSD/fHDHR/fHDHR_web/__init__.py:178} - DEBUG - Client GET requested /api/tuners Closing
[2022-02-19 16:54:58,423] p2990286 {/mnt/ProgSSD/fHDHR/venv/lib/python3.9/site-packages/gevent/pywsgi.py:1226} - INFO - 192.168.2.64 - - [2022-02-19 16:54:58] "GET /api/tuners?method=stream&channel=65089df3-ff53-45b4-8656-9a0844d4edd8&origin=ceton HTTP/1.0" 503 2458 0.712365
```

Thanks!

**Comment by arrmo** [comment]
OK, I may be making a bit of progress with RTP 😜. I have been able to manually control the Ceton (through it's web interface) ... seems like I need to stop the stream (RTP), tune to the channel, then start the stream (thinking fHDHR does the same sequence?). At that point I can capture the video using ffmpeg (vlc still has issues ... arrgh!). But (of course there's a but ... LOL),
1. I can't seem to consistently get fHDHR to do the same thing - this may be me misunderstanding though. In ceton, `__init__.py` I changed to `streamurl = "rtp://192.168.200.2:800%s" % instance` ... but the non-pcie version uses udp, and 127.0.0.1? I may be misunderstanding something here 😞. It doesn't seem to be streaming (consistently ... 1 in 5 seems to work).
2. I do see `bind failed: Address already in use` if I try to use two tuners. This also happens manually, it seems that ffmpeg uses two consecutive addresses (one for the rtcpport, noted [here](https://ffmpeg.org/ffmpeg-protocols.html#rtp)). Still trying to figure out how to change that.

By all means yell if you have any thoughts, or if I'm off base. Still digging!

Thanks.

**Comment by arrmo** [comment]
OK, starting to be able to make it start consistently => if I go to the Ceton UI, force streaming stop, tune, then start ... and then start fHDHR - the first tune works and streams. But when I hit stop (in Plex), I get the error below. Seems this leaves things a bit untidy, so then I have to exit, and manually run the Ceton circle again, LOL!
```
[2022-02-19 19:59:35,609] p3178783 {/mnt/ProgSSD/fHDHR/fHDHR/streammanager/stream/__init__.py:228} - INFO - Stream Ended: Client has disconnected.
[2022-02-19 19:59:35,610] p3178783 {/mnt/ProgSSD/fHDHR/fHDHR/streammanager/stream/__init__.py:235} - INFO - Removing Tuner Lock
[2022-02-19 19:59:35,611] p3178783 {/mnt/ProgSSD/fHDHR/fHDHR/device/tuners/tuner.py:87} - INFO - Tuner #0 Released.
[2022-02-19 19:59:35,612] p3178783 {/mnt/ProgSSD/fHDHR/fHDHR/streammanager/stream/__init__.py:239} - INFO - Removing 3 chunks from the buffer.
Traceback (most recent call last):
  File "/mnt/ProgSSD/fHDHR/venv/lib/python3.9/site-packages/gevent/pywsgi.py", line 946, in run_application
    self.process_result()
  File "/mnt/ProgSSD/fHDHR/venv/lib/python3.9/site-packages/gevent/pywsgi.py", line 932, in process_result
    self.write(data)
  File "/mnt/ProgSSD/fHDHR/venv/lib/python3.9/site-packages/gevent/pywsgi.py", line 775, in write
    self._write(data)
  File "/mnt/ProgSSD/fHDHR/venv/lib/python3.9/site-packages/gevent/pywsgi.py", line 760, in _write
    self._sendall(towrite)
  File "/mnt/ProgSSD/fHDHR/venv/lib/python3.9/site-packages/gevent/pywsgi.py", line 736, in _sendall
    self.socket.sendall(data)
  File "/mnt/ProgSSD/fHDHR/venv/lib/python3.9/site-packages/gevent/_socketcommon.py", line 699, in sendall
    return _sendall(self, data_memory, flags)
  File "/mnt/ProgSSD/fHDHR/venv/lib/python3.9/site-packages/gevent/_socketcommon.py", line 409, in _sendall
    timeleft = __send_chunk(socket, chunk, flags, timeleft, end)
  File "/mnt/ProgSSD/fHDHR/venv/lib/python3.9/site-packages/gevent/_socketcommon.py", line 338, in __send_chunk
    data_sent += socket.send(chunk, flags)
  File "/mnt/ProgSSD/fHDHR/venv/lib/python3.9/site-packages/gevent/_socketcommon.py", line 722, in send
    return self._sock.send(data, flags)
BrokenPipeError: [Errno 32] Broken pipe

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/mnt/ProgSSD/fHDHR/venv/lib/python3.9/site-packages/gevent/pywsgi.py", line 999, in handle_one_response
    self.run_application()
  File "/mnt/ProgSSD/fHDHR/venv/lib/python3.9/site-packages/gevent/pywsgi.py", line 951, in run_application
    close()
  File "/mnt/ProgSSD/fHDHR/venv/lib/python3.9/site-packages/werkzeug/wsgi.py", line 466, in close
    callback()
  File "/mnt/ProgSSD/fHDHR/venv/lib/python3.9/site-packages/werkzeug/wrappers/response.py", line 437, in close
    self.response.close()  # type: ignore
  File "/mnt/ProgSSD/fHDHR/venv/lib/python3.9/site-packages/flask/helpers.py", line 131, in generator
    yield from gen
  File "/mnt/ProgSSD/fHDHR/fHDHR/streammanager/stream/__init__.py", line 243, in buffer_generator
    self.fhdhr.logger.info("Running %s close_stream method." % self.stream_obj.origin)
AttributeError: 'Stream_OBJ' object has no attribute 'origin'
2022-02-20T01:59:35Z {'REMOTE_ADDR': '192.168.2.64', 'REMOTE_PORT': '44144', 'HTTP_HOST': '192.168.2.64:5006', (hidden keys: 24)} failed with AttributeError

[2022-02-19 19:59:35,615] p3178783 {/mnt/ProgSSD/fHDHR/venv/lib/python3.9/site-packages/gevent/pywsgi.py:1226} - INFO - 192.168.2.64 - - [2022-02-19 19:59:35] "GET /api/tuners?method=stream&channel=553&origin=ceton&stream_method=ffmpeg&accessed=http%3A//192.168.2.64%3A5006/hdhr/ceton/auto/v553 HTTP/1.1" 200 42626681 25.970368
```

Thanks!

**Comment by deathbybandaid** [comment]
Check out https://github.com/fHDHR/fHDHR/pull/257

I had to add the origin `@property` to `stream_obj`

**Comment by arrmo** [comment]
Thanks! So close 👍 

I find I still have to tell RTP to start - manually from the Ceton UI. Will look at the code here, but seems like that "start" isn't happening after channel tuning?

**Comment by arrmo** [comment]
And to clarify 😜. If I start streaming (m3u in VLC, Plex, etc.) => I don't change any settings in the Ceton UI, rather ... just go to that tuner, press "Save Streaming Settings" - and away it goes! Perhaps the settings are not being "applied"?

Thanks!

**Comment by arrmo** [comment]
Found it! It's in `plugins/fHDHR_plugin_origin_ceton/origin/__init__.py` ... hacked in for a testing,
```
        StartStop_data = {"instance_id": instance,
                          "dest_ip": '192.168.200.2',
                          #"dest_ip": self.plugin_utils.config.dict["fhdhr"]["address"],
                          "dest_port": 8000 + instance,
                          "protocol": 0,
                          "start": startstop}
```

For the internal card - the destination IP is `192.168.200.2`, and the port is as above ... 8000 + instance_id (i.e. 8000-8003). Any particular way you want this incorporated?

Thanks!

**Comment by arrmo** [comment]
Actually, I'll go ahead and clean this up, submit a PR - by all means comment at that point. Thanks!!

**Comment by arrmo** [comment]
It seems my other (refresh) PR is still pending - will wait on that one, to avoid multiple branches and a bit of a mess 🤣

**Comment by arrmo** [comment]
> the HDHR plugin, the m3u, everything just directs the client (plex, vlc) to `api/tuners?method=stream&channel=XXX`

FYI, I pushed a new PR for the RTP (UDP) code, but when I try to use a channel number, I see this error,
```
[2022-02-23 19:27:27,820] p243570 {/mnt/ProgSSD/fHDHR/fHDHR_web/__init__.py:167} - DEBUG - Client GET requested /api/tuners Opening
Traceback (most recent call last):
  File "/mnt/ProgSSD/fHDHR/venv/lib/python3.9/site-packages/gevent/pywsgi.py", line 999, in handle_one_response
    self.run_application()
  File "/mnt/ProgSSD/fHDHR/venv/lib/python3.9/site-packages/gevent/pywsgi.py", line 945, in run_application
    self.result = self.application(self.environ, self.start_response)
  File "/mnt/ProgSSD/fHDHR/venv/lib/python3.9/site-packages/flask/app.py", line 2076, in wsgi_app
    response = self.handle_exception(e)
  File "/mnt/ProgSSD/fHDHR/venv/lib/python3.9/site-packages/flask/app.py", line 2073, in wsgi_app
    response = self.full_dispatch_request()
  File "/mnt/ProgSSD/fHDHR/venv/lib/python3.9/site-packages/flask/app.py", line 1518, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/mnt/ProgSSD/fHDHR/venv/lib/python3.9/site-packages/flask/app.py", line 1516, in full_dispatch_request
    rv = self.dispatch_request()
  File "/mnt/ProgSSD/fHDHR/venv/lib/python3.9/site-packages/flask/app.py", line 1502, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
  File "/mnt/ProgSSD/fHDHR/fHDHR_web/api/tuners.py", line 20, in __call__
    return self.get(*args)
  File "/mnt/ProgSSD/fHDHR/fHDHR_web/api/tuners.py", line 32, in get
    stream_obj, stream_args, response_dict = streammanager.setup_stream(request, session)
  File "/mnt/ProgSSD/fHDHR/fHDHR/streammanager/__init__.py", line 81, in setup_stream
    chan_obj, response_dict = self.parse_channel_args_for_chan_obj(request)
  File "/mnt/ProgSSD/fHDHR/fHDHR/streammanager/__init__.py", line 118, in parse_channel_args_for_chan_obj
    chan_obj = self.device.channels.get_channel_obj_search(origin, channel_number)
  File "/mnt/ProgSSD/fHDHR/fHDHR/device/channels/__init__.py", line 63, in get_channel_obj_search
    return chan_obj
UnboundLocalError: local variable 'chan_obj' referenced before assignment
2022-02-24T01:27:27Z {'REMOTE_ADDR': '192.168.2.66', 'REMOTE_PORT': '49816', 'HTTP_HOST': 'server:5006', (hidden keys: 27)} failed with UnboundLocalError
[2022-02-23 19:27:27,825] p243570 {/mnt/ProgSSD/fHDHR/venv/lib/python3.9/site-packages/gevent/pywsgi.py:1226} - INFO - 192.168.2.66 - - [2022-02-23 19:27:27] "GET /api/tuners?method=stream&channel=593 HTTP/1.1" 500 161 0.009017
```

**Comment by arrmo** [comment]
Arrgh 😜. OK, stumbled onto this ... it's not direct vs. RTP related. I was running RTP last night, and had the same lagging / video stutter issue - turns out, it was because I had also changed the logging to debug. That's the trigger! So I ran (all this with m3u, direct to VLC) ...
1. RTP, debug = info: 90 minutes, 325k frames, 6 lost video frames, no input dropped / discontinued
2. RTP, debug = debug: starts OK, gets worse with time, after ~ 40 min or so (it restarted, so not exact), 178 lost frames, 5 input dropped / disconnected
3. Back to RTP, debug = info ... very similar to 1 above

It's not really just CPU load though either, as it progressively gets worse. Memory leak of some sort?

I want to go back, try direct again, but with the latest streaming updates I now have this issue with direct?
```
[2022-02-24 14:15:01,976] p411428 {/mnt/ProgSSD/fHDHR/plugins/fHDHR_plugin_origin_ceton/origin/__init__.py:219} - INFO - Initiate PCIe direct streaming, channel 593 from Ceton tuner#: 0
[2022-02-24 14:15:01,977] p411428 {/mnt/ProgSSD/fHDHR/plugins/fHDHR_plugin_origin_ceton/origin/__init__.py:224} - INFO - Ceton tuner 0 streamurl set, to: /dev/ceton/ctn91xx_mpeg0_0
[2022-02-24 14:15:01,978] p411428 {/mnt/ProgSSD/fHDHR/fHDHR/streammanager/stream_obj.py:98} - INFO - Attempting to Select an available tuner for this stream.
[2022-02-24 14:15:01,978] p411428 {/mnt/ProgSSD/fHDHR/fHDHR/device/tuners/tuner.py:76} - INFO - Tuner #0 Acquired.
[2022-02-24 14:15:01,979] p411428 {/mnt/ProgSSD/fHDHR/fHDHR/streammanager/stream_obj.py:106} - INFO - ceton Tuner #0 to be used for stream.
[2022-02-24 14:15:01,980] p411428 {/mnt/ProgSSD/fHDHR/fHDHR/streammanager/stream_obj.py:108} - INFO - Preparing Stream...
[2022-02-24 14:15:01,980] p411428 {/mnt/ProgSSD/fHDHR/fHDHR/streammanager/stream/__init__.py:61} - INFO - Stream Method Detected as a /dev/ hardware device.
[2022-02-24 14:15:01,981] p411428 {/mnt/ProgSSD/fHDHR/fHDHR/streammanager/stream_obj.py:72} - INFO - Client has requested stream for ceton channel 593 Golf HD Golf HD.
[2022-02-24 14:15:01,982] p411428 {/mnt/ProgSSD/fHDHR/fHDHR/streammanager/stream_obj.py:78} - INFO - Selected Stream Parameters:method=direct duration=0 origin_quality=None transcode_quality=None.
[2022-02-24 14:15:01,982] p411428 {/mnt/ProgSSD/fHDHR/fHDHR/streammanager/stream/__init__.py:111} - INFO - Tuning Stream...
[2022-02-24 14:15:01,983] p411428 {/mnt/ProgSSD/fHDHR/fHDHR_web/__init__.py:181} - DEBUG - Client GET requested /api/tuners Closing
[2022-02-24 14:15:01,987] p411428 {/mnt/ProgSSD/fHDHR/fHDHR/streammanager/stream/direct_hardware_stream.py:27} - INFO - Direct Hardware Stream from device: /dev/ceton/ctn91xx_mpeg0_0
[2022-02-24 14:15:01,987] p411428 {/mnt/ProgSSD/fHDHR/fHDHR/streammanager/stream/direct_hardware_stream.py:36} - DEBUG - Operating System is: Linux
[2022-02-24 14:15:01,987] p411428 {/mnt/ProgSSD/fHDHR/fHDHR/streammanager/stream/direct_hardware_stream.py:37} - DEBUG - Command is: cat /dev/ceton/ctn91xx_mpeg0_0
cat: /dev/ceton/ctn91xx_mpeg0_0: Device or resource busy
[2022-02-24 14:15:02,003] p411428 {/mnt/ProgSSD/fHDHR/fHDHR/streammanager/stream/direct_hardware_stream.py:46} - DEBUG - Pulling Chunk #1
[2022-02-24 14:15:02,005] p411428 {/mnt/ProgSSD/fHDHR/fHDHR/streammanager/stream/direct_hardware_stream.py:27} - INFO - Direct Hardware Stream from device: /dev/ceton/ctn91xx_mpeg0_0
[2022-02-24 14:15:02,006] p411428 {/mnt/ProgSSD/fHDHR/fHDHR/streammanager/stream/direct_hardware_stream.py:36} - DEBUG - Operating System is: Linux
[2022-02-24 14:15:02,007] p411428 {/mnt/ProgSSD/fHDHR/fHDHR/streammanager/stream/direct_hardware_stream.py:37} - DEBUG - Command is: cat /dev/ceton/ctn91xx_mpeg0_0
cat: /dev/ceton/ctn91xx_mpeg0_0: Device or resource busy
```
If I exit, manually set up the tuner, and stream from the device to mplayer ... all good!

Thanks!

**Comment by arrmo** [comment]
OK, figured out the direct playback issue! I'll note it in the Ceton PR I have open - but the issue comes down to ... you can use direct, or RTP, not both (i.e. they are mutually exclusive). So StartStop can't be called, if using direct. And of course ... why didn't it happen before? Because the StartStop address (and port) was incorrect => Ceton was ignoring it. Once I corrected that - then you see the limit (one or the other, kinda makes sense). Oh, and with a fix (local yet), direct works fine - no lagging or issues, as long as debug = info.

Give your other PR for channels - I think all that is above is related to logging ... lagging for debug (or higher), streaming fail for ssdp. Want me to keep those here, or create a new (or 2?) Issue? Whatever you prefer.

Thanks!

**Comment by arrmo** [comment]
OK, I created a separate issue for the Tuner item (failing when certain log level is set) ... keep this one for the topic the title captures => dropped video. The issue is as noted above, seeing `TS discontinuity` errors - and it seems to be log level dependent (worse with `debug`, vs. `info`), and also gradually gets worse vs. time. A few things I can think to try ... suggestions?
1. turn buffering off in fHDHR. Just to see if it has an impact. We had discussed this, I just need go go back, confirm what variable to set.
2. try changing the buffer size, make it an integer number of mpeg-ts packets (3984 bytes x N)
3. profile the code while running, see if CPU load or memory utilization changes. Not sure this one is so easy though.

I THINK that just restarting the player "fixes" (restarts) the issue - but need to confirm that. That may help point to where it's happening.

Thanks!
